### PR TITLE
Add time extraction functions to project time as a dimension

### PR DESCRIPTION
**Pull request by xvrl**
Time extraction functions are similar to dimension extraction functions in that they allow mapping the time column to arbitrary Strings. Unlike granularities, they allow for non-linear mapping and aggregation across the resulting values. For example, this allows aggregating by day of the week or hour of the day and return the results either as part of a topN or groupBy similar to any other dimension.

This pull request proposes the addition of TimeExtractionDimensionSpec as a new type of DimensionSpec, as well as a the TimeExtractionFn type.

Time is not yet currently available as a true dimension, and this is not the objective of this pull request.
Current topN and groupBy queries still heavily rely on having dimension value indices, which makes it difficult to operate on an arbitrary column. For the sake of cleaner code, it therefore felt more natural to leave the concept of time extraction separate from the existing dimension extraction. This also allows for specific optimizations in for topN and groupBy, given the properties of the time dimension.

Hopefully this can help shape future work towards supporting arbitrary columns as dimensions.

### CODE REVIEW

**Comment by cheddar** [comment]
I just looked at the high level implementation at this point, but I'm a little worried that this is just a further step towards having an extraction interface for every possible primitive type.  The fact that this requires adding a new method which is essentially just specializing on a primitive type makes me wonder if the initial "extraction fn" abstraction was correct.

My first high level thought is wondering if we can invert it so that we pass in something that allows the "extraction function" to get the type of value that it wants.  Kinda like the MetricColumnSelectors.  I can't claim to have thought it through all the way, but the thing that immediately jumps out as less than wonderful in what I'm proposing is that it would complicate implementations by making them care about multi-valued columns.  It would also likely negate some of the "caching" of extraction values that can be done with dimensions...

Another option could be changing the interface to take in an Object instead of String.  That would allow an implementation to be passed a `long` instead of a `String`.  Given that we would likely implement these functions on top of JodaTime (doing `new DateTime()`) this would actually enable one function to handle both `String` dimensions and `long` timestamp type things.  It would introduce object creation overhead when using the Time style ones, but maybe that's not the end of the world?

Anyway, I have to run now.  I'm not necessarily done thinking, but going to hit comment so that the seeds of these thoughts can be planted.

**Comment by xvrl** [comment]
For a lower-level interface I don't necessarily think that having different function methods for different primitive types is necessarily a bad thing, or at least not worse than having to do instance type checks within a method taking Objects. Hopefully Java will provide value types one day and that won't be a problem anymore. We can offer a higher-level interface for simpler functions and implement various casting operations for convenience, leaving the option to implement the low-level interface directly. Low-level functions could expose capabilities to define the supported types, which can in turn can be leveraged by query engines to optimize accordingly.

Given that the type of functions we are talking about operate on single values and not on rows, I don't think there is an immediate need to invert it. If that is what we nonetheless decide to do in the future, we can easily wrap existing functions to use the inversion mechanisms.

More immediately, I believe the most important thing to focus on is the interface to the user, so we don't have to change the query language too much. Maybe it makes more sense to provide a general `ExtractionDimensionSpec`, introducing a new `extractionFn` field and start deprecating the `dimExtractionFn` field. The result should still be thought of as a dimension though. That would remove the time-specific aspect from the query interface, and leave it to the query engine whether to support non-string values for columns other than time for now. This also opens up the ability to use all the existing dimExtraction Functions on the time column.

File: processing/src/main/java/io/druid/query/topn/TopNParams.java [hunk 1]
```diff
@@ -29,6 +29,13 @@
   private final int cardinality;
   private final int numValuesPerPass;
 
+  /**
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNParams.java]: Please either populate or remove.

File: processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import io.druid.query.extraction.TimeExtractionFn;
+import io.druid.segment.Capabilities;
+import io.druid.segment.Cursor;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.column.Column;
+
+import java.util.Map;
+
+public class TimeExtractionTopNAlgorithm extends BaseTopNAlgorithm<int[], Map<String, Aggregator[]>, TopNParams>
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java]: Do you have any performance numbers on this vs pooled vs DimExtr topN?

**Comment by drcrallen** [comment]
I agree with @xvrl on the low level aspect.

In a general sense, I think anything people may want to do can be accomplished with a javascript extraction function, but we are supplying optimized extractions as they become major use cases. 

I think eventually we want the `__time` dimension to be just another dimension, but that's a bit too big of an ask for this PR.

Rather than `ExtractionDimensionSpec` how would you feel about `EphemeralDimensionSpec`? Specifically for a dimension that is only in existence for the life of the query. (also might be too much scope creep for this PR)

**Comment by xvrl** [comment]
I reworked the interfaces a have a more generic ExtractionFn, with DimExtractionFn being a special case for String dimensions. This allows us to use any existing DimExtraction function on the time column, and I generalized JavaScriptExtractionFunction to not be specific to Strings anymore.

Would this seem more palatable to everyone?

File: processing/src/main/java/io/druid/query/extraction/DimExtractionFn.java [hunk 3]
```diff
@@ -1,73 +1,35 @@
...
-   *
-   *
-   * @param dimValue the original value of the dimension
-   * @return a value that should be used instead of the original
-   */
-  public String apply(String dimValue);
+  @Override
+  public String apply(Object value)
+  {
+    return apply(value.toString());
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/extraction/DimExtractionFn.java]: Does the `toString()` increase query time for String based extraction functions?

Comment (xvrl) [inline processing/src/main/java/io/druid/query/extraction/DimExtractionFn.java]: toString() is just a default implementation, it is not called for String dimensions, those use apply(String) directly as it did before.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 4]
```diff
@@ -312,12 +314,99 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+                    @Override
+                    public int size()
+                    {
+                      return 1;
+                    }
+
+                    @Override
+                    public int get(int index)
+                    {
+                      return getTimeBucketIndex();
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Can you explain this a little? Is it assuming that get(index) will always ask for sequential index values?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 5]
```diff
@@ -312,12 +314,99 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+                {
+                  throw new UnsupportedOperationException("time column does not support lookups");
+                }
+              }
+          );
           dimNames.add(dimSpec.getOutputName());
+        } else {
+          final DimensionSelector selector = cursor.makeDimensionSelector(dimension);
+          if (selector != null) {
+            dimensions.add(selector);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: dimName is not added if using time extraction.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 6]
```diff
@@ -312,12 +314,99 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+                          timeValues.put(index, currentValue);
+                        }
+                      }
+                      return index;
+                    }
+
+                    @Override
+                    public Iterator<Integer> iterator()
+                    {
+                      return Iterators.forArray(getTimeBucketIndex());
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Also that get(index) and iterator() will never be called on the same object?

File: processing/src/main/java/io/druid/query/topn/TopNQuery.java [hunk 7]
```diff
@@ -139,7 +139,7 @@ public QueryGranularity getGranularity()
 
   public void initTopNAlgorithmSelector(TopNAlgorithmSelector selector)
   {
-    if (dimensionSpec.getDimExtractionFn() != null) {
+    if (dimensionSpec.getExtractionFn() != null) {
       selector.setHasDimExtractionFn(true);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNQuery.java]: May as well rename this to keep future confusion at a minimum.

File: processing/src/main/java/io/druid/query/extraction/DimExtractionFn.java (hunk 3 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/extraction/DimExtractionFn.java]: nvmnd, I see the String override from the inheritance.

File: processing/src/main/java/io/druid/query/extraction/JavascriptExtractionFn.java [hunk 8]
```diff
@@ -90,10 +88,21 @@ public String getFunction()
...
+  {
+    return this.apply((Object)value);
+  }
+
+  @Override
+  public String apply(long value)
   {
-    String retVal = fn.apply(dimValue);
-    return Strings.isNullOrEmpty(retVal) ? null : retVal;
+    return this.apply(Long.valueOf(value));
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/extraction/JavascriptExtractionFn.java]: I wonder if there's not a reason to just let it auto-box?

Comment (xvrl) [inline processing/src/main/java/io/druid/query/extraction/JavascriptExtractionFn.java]: good point

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 9]
```diff
@@ -312,12 +314,99 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+
+                    @Override
+                    public int get(int index)
+                    {
+                      return getTimeBucketIndex();
+                    }
+
+                    protected Integer getTimeBucketIndex()
+                    {
+                      final long timestamp = timeSelector.get();
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: I think you can move this logic to exist before `return new IndexedInts()...` and effectively just generate an `final int index` which is passed into the anonymous class.

Hopefully that will clarify what is happening in the code as well.

File: processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java (hunk 2 above)
Comment (cheddar) [inline processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java]: Fwiw, I believe this is known to be slower, but it is a necessity unless we actually "materialize" all of the possible outputs upfront and treat that as a dimension.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 4 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Yes, this assumes we are scanning through the segment once from start to end, and that timestamps are grouped. I will add more comments to make that clear.

**Comment by cheddar** [comment]
I like the changes, moving to just an `ExtractionFn` and giving that interface various methods that enable specialization for specific primitives makes sense to me.

File: processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java (hunk 2 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java]: Pooled topN assumes relies on dimension cardinality to do multiple passes, which is not available for time. The time-extraction specific algorithm is essentially the dimension extraction algorithm, modified to bypass the need for `DimSelector`, so it should not be any slower than dimension extraction topN.

File: processing/src/main/java/io/druid/query/topn/TopNQuery.java (hunk 7 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TopNQuery.java]: indeed, I missed that one

**Comment by xvrl** [comment]
@cheddar @drcrallen addressed your comments and added documentation.

**Comment by drcrallen** [comment]
I'm going to have a lot of cleanup to merge this into QTL but this looks good to me

**Comment by vogievetsky** [comment]
Very excited to see this feature. Just updated the TypeScript types in anticipation: https://github.com/facetjs/typescript-druid/commit/5a0778b1d14a085b3f43534650e83d8d4199134d

BTW. It would be great to add a full query example on the doc page.

**Comment by xvrl** [comment]
@vogievetsky I added an example [here](https://github.com/druid-io/druid/pull/1159/files#diff-40f1c0477ed1df59af6953a37656d768R98)

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 10]
```diff
@@ -312,12 +314,113 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
       dimensions = Lists.newArrayListWithExpectedSize(dimensionSpecs.size());
       dimNames = Lists.newArrayListWithExpectedSize(dimensionSpecs.size());
+
       for (int i = 0; i < dimensionSpecs.size(); ++i) {
         final DimensionSpec dimSpec = dimensionSpecs.get(i);
-        final DimensionSelector selector = cursor.makeDimensionSelector(dimSpec.getDimension());
-        if (selector != null) {
-          dimensions.add(selector);
+        final String dimension = dimSpec.getDimension();
+        if(dimension.equals(Column.TIME_COLUMN_NAME)) {
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: I'm wondering if this logic shouldn't be given to the cursor and done inside of `makeDimensionSelector`.  I'm assuming that you chose not to do it that way because that means each implementation of `StorageAdapter` has to implement the time lookup properly, but it also seems like each implementation of `StorageAdapter` might be able to take advantage of optimizations when returning the time column...

I guess it seems like we are building in extra logic about "the time column acts like XYZ", where I think that with 0.7 there was an effort to make the time column just look like any other column.  This seems to be breaking that "similarity"...

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 11]
```diff
@@ -395,11 +498,20 @@ public Row apply(@Nullable Map.Entry<ByteBuffer, Integer> input)
...
-                    final DimExtractionFn fn = dimensionSpecs.get(i).getDimExtractionFn();
+                    final DimensionSpec dimensionSpec = dimensionSpecs.get(i);
+                    final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
                     final int dimVal = keyBuffer.getInt();
                     if (dimSelector.getValueCardinality() != dimVal) {
-                      if (fn != null) {
-                        theEvent.put(dimNames.get(i), fn.apply(dimSelector.lookupName(dimVal)));
+                      if (extractionFn != null) {
+                        final String value;
+                        // extraction functions are applied at the dimSelector level for time
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Can you explain this optimization a bit more?  Also, have you benchmarked if it actually helps?  It seems like it is adding extra branch complexity in the code, which can sometimes be counter productive.

File: processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java [hunk 12]
```diff
@@ -88,7 +89,15 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
   private Function<Cursor, Result<TopNResultValue>> getMapFn(TopNQuery query, final StorageAdapter adapter)
   {
     final Capabilities capabilities = adapter.getCapabilities();
-    final int cardinality = adapter.getDimensionCardinality(query.getDimensionSpec().getDimension());
+    final String dimension = query.getDimensionSpec().getDimension();
+
+    final int cardinality;
+    if (dimension.equals(Column.TIME_COLUMN_NAME)) {
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java]: This seems weird...  Can the adapter take care of it instead?

File: processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java [hunk 13]
```diff
@@ -97,8 +106,10 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
     final TopNAlgorithmSelector selector = new TopNAlgorithmSelector(cardinality, numBytesPerRecord);
     query.initTopNAlgorithmSelector(selector);
 
-    TopNAlgorithm topNAlgorithm = null;
-    if (selector.isHasDimExtractionFn()) {
+    final TopNAlgorithm topNAlgorithm;
+    if (selector.isHasExtractionFn() && dimension.equals(Column.TIME_COLUMN_NAME)) {
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java]: The `Column.TIME_COLUMN_NAME` is getting proliferated throughout the code quite a bit.  I think we should try to contain it as much as possible.

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    @JsonSubTypes.Type(name = "timeFormat", value = TimeFormatExtractionFn.class)
+})
+/**
+ * An ExtractionFn is a function that can be used to transform the values of a column (typically a dimension)
+ *
+ * A simple example of the type of operation this enables is the RegexDimExtractionFn which applies a
+ * regular expression with a capture group.  When the regular expression matches the value of a dimension,
+ * the value captured by the group is used for grouping operations instead of the dimension value.
+ */
+public interface ExtractionFn
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: Just double checked the tests and it looks like we don't have serde tests that will verify backwards compatibility for us.  How do you feel about adding some for the various implementations and paying down a bit of tech debt?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 11 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: not optimizing for time would mean that consecutive (in scan order), but different time values that map to the same resulting string (e.g. `Monday`) would each have to have a different `dimVal` index instead of sharing the same one, because the mapping cannot be used earlier when it actually scans the column (https://github.com/druid-io/druid/pull/1159/files#diff-7d67ab242d530f3d34f752babacd7090L364)

It would also mean that dimSelector.lookupName() would need to support returning long values or Object instead of just String in order to not incur long serialization, which is a much bigger undertaking.

File: processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java (hunk 13 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java]: Agree, they're not pretty. In this case `dimension.equals(Column.TIME_COLUMN_NAME)` really means "is this a dimension of type long and of unknown cardinality". I'm not sure if there is a better way to express that in the current state, but I can add a comment.

File: docs/content/DimensionSpecs.md [hunk 15]
```diff
@@ -16,63 +18,139 @@ Returns dimension values as is and optionally renames the dimension.
...
 {
   "type" : "extraction",
   "dimension" : <dimension>,
   "outputName" :  <output_name>,
-  "dimExtractionFn" : <dim_extraction_fn>
+  "extractionFn" : <extraction_function>
 }
 ```
 
-## <a id="toc_3"></a>DimExtractionFn
```
Comment (fjy) [inline docs/content/DimensionSpecs.md]: was this hyperlink not used?

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java (hunk 14 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: +1

File: processing/src/main/java/io/druid/query/extraction/TimeFormatExtractionFn.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+      @JsonProperty("locale") String localeString
+  )
+  {
+    Preconditions.checkArgument(pattern != null, "format cannot be null");
+
+    this.pattern = pattern;
+    this.tz = tz;
+    this.locale = localeString == null ? null : Locale.forLanguageTag(localeString);
+    this.formatter = DateTimeFormat.forPattern(pattern)
+                                   .withZone(tz == null ? DateTimeZone.UTC : tz)
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/extraction/TimeFormatExtractionFn.java]: can we define the default timezone as a static constant?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 10 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: I agree that is logic seems out of place here. We should try to centralize the places that have to special case for the time column

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 17]
```diff
@@ -312,12 +314,113 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+                private String currentValue = null;
+                private long currentTimestamp = Long.MIN_VALUE;
+                private int index = -1;
+
+                @Override
+                public IndexedInts getRow()
+                {
+                  // if this the first timestamp, apply and cache extraction function result
+                  final long timestamp = timeSelector.get();
+                  if(index < 0) {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: there's some formatting problems in the PR

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 18]
```diff
@@ -312,12 +314,113 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
...
+                @Override
+                public int lookupId(String name)
+                {
+                  throw new UnsupportedOperationException("time column does not support lookups");
+                }
+              }
+          );
           dimNames.add(dimSpec.getOutputName());
+        } else {
+          final DimensionSelector selector = cursor.makeDimensionSelector(dimension);
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: can this be outside the if/else loop?

File: processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+
+  @Override
+  public TopNParams makeInitParams(DimensionSelector dimSelector, Cursor cursor)
+  {
+    return new TopNParams(
+        dimSelector,
+        cursor,
+        Integer.MAX_VALUE, // artificially set cardinality high to force a single pass
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java]: this seems hacky, can we not force a single pass through other means?

File: docs/content/DimensionSpecs.md (hunk 15 above)
Comment (xvrl) [inline docs/content/DimensionSpecs.md]: links are created automatically now

File: processing/src/main/java/io/druid/query/extraction/TimeFormatExtractionFn.java (hunk 16 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/extraction/TimeFormatExtractionFn.java]: not sure what you mean, DateTimeZone.UTC is already a static constant.

File: processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java (hunk 19 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TimeExtractionTopNAlgorithm.java]: TopN is currently deeply tied to DimensionSelector, so this would require a major rewrite of topN, which I would rather leave to when we decide to support arbitrary columns as dimensions.

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java (hunk 14 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: There were no serde changes at the ExtractionFn level, we just renamed the interface and added some methods. The backwards compatibility is at the ExtractionDimensionSpec level, so I'm happy to add serde tests there.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 18 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: not if makeDimensionSelector does not work for `__time`

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 10 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: I'm not sure which one is better frankly. The query engine has better assumptions regarding how the data is scanned, so it can potentially make better decisions on how to optimize, based on the storage adapter capabilities. The Storage adapter may be able to optimize certain things, but would have to handle the more general case, since it doesn't know what the query engine might do.

We can push down all extraction functions to the storage adapter level, if you feel this would be more consistent?

File: processing/src/main/java/io/druid/query/extraction/ExtractionFn.java (hunk 14 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/extraction/ExtractionFn.java]: ok, I added a bunch of serde tests for extraction functions, and extraction dimension spec backwards compatibility.

**Comment by xvrl** [comment]
@cheddar I decided to push down extraction functions to the storage adapter level.
- The code is now much cleaner, with fewer cases to handle for extraction functions in the query engines.
- topN with timeExtraction is now twice as fast, thanks to the added id to value map in the dimselector.
- This also opens up the possibility to cache extraction function results at the storage adapter level, similar to what we can do with actual dimension values.

The only change in behavior this introduces, is that when specifying a lexicographic topn metric spec with a dimension extraction function, you could specify the actual dimension value (before applying the extraction function) as "last stop" to optimize. This optimization is not possible anymore, but in my opinion irrelevant, since that would previously require knowledge of the underlying values, or have the prior query use a dimension extraction function that did not modify the values.

**Comment by cheddar** [comment]
Cool.  I haven't really had time to go into the details of the change, but I'm happy with an initial cursory glance over what's there.  So I'm generally +1

File: processing/src/main/java/io/druid/query/search/SearchQueryRunner.java [hunk 20]
```diff
@@ -161,7 +161,7 @@ public SearchQueryRunner(Segment segment)
 
             Map<String, DimensionSelector> dimSelectors = Maps.newHashMap();
             for (String dim : dimsToSearch) {
-              dimSelectors.put(dim, cursor.makeDimensionSelector(dim));
+              dimSelectors.put(dim, cursor.makeDimensionSelector(dim, null));
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/search/SearchQueryRunner.java]: Can makeDimensionSelector simply take a dim argument and wrap `.makeDimensionSelector(dim, null)` (no extraction function) instead of having to have null all over the place?

Just looking at this function without knowing about the dim extraction function parameter makes the operation of the method less obvious.

File: processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java [hunk 21]
```diff
@@ -88,7 +89,11 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
   private Function<Cursor, Result<TopNResultValue>> getMapFn(TopNQuery query, final StorageAdapter adapter)
   {
     final Capabilities capabilities = adapter.getCapabilities();
-    final int cardinality = adapter.getDimensionCardinality(query.getDimensionSpec().getDimension());
+    final String dimension = query.getDimensionSpec().getDimension();
+
+    final int cardinality;
+    cardinality = adapter.getDimensionCardinality(dimension);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java]: I assume there was something else here at one time that warranted a line break

Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java]: just an artifact of changing the code back and forth, will fix

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java [hunk 22]
```diff
@@ -308,13 +315,17 @@ public int getValueCardinality()
                           @Override
                           public String lookupName(int id)
                           {
-                            final String retVal = column.lookupName(id);
-                            return retVal == null ? "" : retVal;
+                            final String value = column.lookupName(id);
+                            final String retVal = extractionFn == null ? value : extractionFn.apply(value);
+                            return Strings.nullToEmpty(retVal);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: This breaks convention in a lot of other places. Most places I can recall translate "" to null, not the other way around.

File: processing/src/main/java/io/druid/query/search/SearchQueryRunner.java (hunk 20 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/search/SearchQueryRunner.java]: I think it might make more sense to update Search and Select queries to support proper dimensionSpecs with extraction as opposed to just simple dimensions, in which case we would not need the single argument method.

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java [hunk 23]
```diff
@@ -272,14 +275,18 @@ public void reset()
                     }
 
                     @Override
-                    public DimensionSelector makeDimensionSelector(String dimension)
+                    public DimensionSelector makeDimensionSelector(final String dimension, @Nullable final ExtractionFn extractionFn)
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: Do you have some benchmarks on these changes?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 24]
```diff
@@ -209,12 +210,14 @@ public int getValueCardinality()
...
           {
-            return in.get().getDimension(dimension).get(id);
+            final String value = in.get().getDimension(dimension).get(id);
+            return extractionFn == null ? value : extractionFn.apply(value);
           }
 
           @Override
           public int lookupId(String name)
           {
+            if(extractionFn == null) throw new UnsupportedOperationException("cannot perform lookup when applying an extraction function");
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I think this logic is reversed.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 25]
```diff
@@ -163,7 +164,7 @@ public Object get()
       }
 
       @Override
-      public DimensionSelector makeDimensionSelector(final String dimension)
+      public DimensionSelector makeDimensionSelector(final String dimension, final ExtractionFn extractionFn)
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: Same question as before, do you have some benchmarks to check the performance impact of these changes?

Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: What is your concern, specifically?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 24 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: Indeed

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java (hunk 22 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: Ah I might have inadvertantly changed the behavior here, but I will leave the nulls to empty before applying the extraction, since that's what it did before.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 25 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: column selector is at the heart of lots of calls. Even though the lines added to the logic are very small, they will probably get called many many many times in potentially performance critical paths. If it weren't in such a critical portion I wouldn't be concerned. I'm most concerned that before and after the change for things like topN or groupBy do not change. DimExtractionTopN simply performs worse in general, but it would be good to know if this its performance changes much with the column selector changes.

File: processing/src/main/java/io/druid/query/search/SearchQueryRunner.java (hunk 20 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/search/SearchQueryRunner.java]: Sounds good. I'm ok preventing that from being scope creep for this PR, but would you mind adding a TODO note here indicating that desire?

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/search/SearchQueryRunner.java]: *And the other places it applies

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 25 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: For topN and groupBy the actual lookups are performed only after doing the aggregation, and in those cases we already did the check for extraction function as part of the query engine before performing the lookup. Now it is performed within the selector instead of outside, so this should not affect performance for typical queries. 

The only queries that might be affected are aggregators that do the lookup as part of the object column selector to work on actual values, such as JavaScript and Cardinality. But in those instances a null  check is very cheap compared to the aggregation, and even then branch prediction will likely remove that overhead.
